source/upyun/upyun_test.go

```diff
@@ -0,0 +1,171 @@
+// +-------------------------------------------------------------------------
+// | Copyright (C) 2016 Yunify, Inc.
+// +-------------------------------------------------------------------------
+// | Licensed under the Apache License, Version 2.0 (the "License");
+// | you may not use this work except in compliance with the License.
+// | You may obtain a copy of the License in the LICENSE file, or at:
+// |
+// | http://www.apache.org/licenses/LICENSE-2.0
+// |
+// | Unless required by applicable law or agreed to in writing, software
+// | distributed under the License is distributed on an "AS IS" BASIS,
+// | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// | See the License for the specific language governing permissions and
+// | limitations under the License.
+// +-------------------------------------------------------------------------
+
+package upyun
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/upyun/go-sdk/upyun"
+	"github.com/yunify/qscamel/record"
+	"github.com/yunify/qscamel/utils"
+)
+
+const (
+	envBucketName = "QSCAMEL_TEST_UPYUN_BUCKETNAME"
+	envZone       = "QSCAMEL_TEST_UPYUN_ZONE"
+	envAccessKey  = "QSCAMEL_TEST_UPYUN_ACCESS_KEY_ID"
+	envSecretKey  = "QSCAMEL_TEST_UPYUN_SECRET_ACCESS_KEY"
+)
+
+type testCase struct {
+	SourceObjects       []string
+	ThreadNum           int
+	RecordObjects       []string
+	ExpectedObjectNames []string
+	ExpectedSkipped     []string
+	ExpectedDone        bool
+}
+
+var sourceUpyunTests = []testCase{
+	// Normal
+
+	{
+		SourceObjects:       []string{"object1", "object2", "object3"},
+		ThreadNum:           3,
+		ExpectedObjectNames: []string{"object1", "object2", "object3"},
+		// IsTruncated is false when limit(maxKeys) is exactly equal to object num.
+		ExpectedDone: false,
+	},
+
+	// TreadNum > total number of objects
+	{
+		SourceObjects:       []string{"object4", "object5", "object6"},
+		ThreadNum:           5,
+		ExpectedObjectNames: []string{"object4", "object5", "object6"},
+		ExpectedDone:        true,
+	},
+
+	// Skip recorded objects
+	{
+		SourceObjects:       []string{"object7", "object8", "object9"},
+		ThreadNum:           3,
+		RecordObjects:       []string{"object7"},
+		ExpectedObjectNames: []string{"object8", "object9"},
+		ExpectedSkipped:     []string{"object7"},
+		ExpectedDone:        true,
+	},
+}
+
+func TestSourceUpyun_GetSourceSites(t *testing.T) {
```

Comment:
You mean the test function name should remove “_” ?   If so, there are a lot of functions should be changed...